### ISSUE TITLE
Circular deps graph is unproportionally huge

### ISSUE THREAD

**Issue by markomitranic** [symfony/symfony#37540]
**Symfony version(s) affected**: 3.4.42

**Description**  
My team is experiencing huge issues with container built times on Symfony 3.4.42. We are working heavily on performing the update to 4.4 (which does not exhibit the same issue). We have observed this issue on earlier versions, and can track them back to 3.4.37, before which we have no reliable data.

But thisis an interesting sight, and it would mean quite a lot if i could get @nicolas-grekas thoughts on this particular [Blackfire profile](https://blackfire.io/profiles/6f371d70-e697-4cc9-b1dc-0fc4c3621281/graph?settings%5Bdimension%5D=wt&settings%5Bdisplay%5D=landscape&settings%5BtabPane%5D=nodes&selected=&callname=main()). < This is a cold start, a first pageload of an application without var/cache/*...

As you might imagine, 29 million method calls takes quite a lot of time, and prevents us from using xdebug at all (its not installed in the container that was used to make the abovementioned profile), or often makes blackfire probe unavailable because of timeouts. We do have about ~1500 manually registered services, most of which use autowiring and autoconfiguration, and we do not let the app scan for services automagically.

Any kind of hint would help. <3

**How to reproduce**  
This is very tough to reproduce as i have absolutely no idea what the minimum viable code amount would be, as the problem seems thightly scaling with the sheer number of autowired services (1693).

If anyone has any ideas on how i can make a reproducer, feel free to tell me. For reference here are a few examples of what may be found in our `services.yml`

```
    App\CoreBundle\DependencyInjection\MetaTagsService:
        autoconfigure: true
        autowire: true

    App\CoreBundle\DependencyInjection\CustomListService:
        autoconfigure: true
        autowire: true
        arguments:
            - '@service_container'
            - '@doctrine.odm.mongodb.document_manager'
            - '@geo_target'
            - '@translator'
        tags:
            - { name: doctrine_mongodb.odm.event_listener, event: prePersist }

    redirection_service:
        class: App\CoreBundle\DependencyInjection\RedirectionService
        arguments:
            - '@activity_log_service'
        tags:
            - { name: doctrine_mongodb.odm.event_listener, event: prePersist }
            - { name: doctrine_mongodb.odm.event_listener, event: preUpdate }
            - { name: doctrine_mongodb.odm.event_listener, event: postFlush }

    App\CoreBundle\Service\AgCash\AgCoinStatusService:
        autowire: true
        autoconfigure: true
        arguments:
            $transformer: '@App\CoreBundle\Service\Transformer\API\General\QuickSearch\QuickSearchOutputTransformer'
            $translator: '@translator'
            $authorizationChecker: '@security.authorization_checker'
            $currentCountryProvider: '@current_country_provider'
```

**Possible Solution**  
Not yet. :( 

**Additional context**  
The BF profile is clearly pointing to 3 methods that perform 29M loops, but dumping the variables inside the methods shows nothing unexpected. $this->circularReferences is at count 168, total number of services is 1693.

My best guess was that this is related to the fix in #29336 and the fact that @nicolas-grekas seem to have rolled back on some of the changes as of [this commit](https://github.com/symfony/symfony/commit/ed96830893ca5c7052cf4aee08591a9ec4b0c571) because of compatibility reasons. But, i am by no means an expert on the topic, and need help interpreting some of the changes made in those two.

**Comment by nicolas-grekas** [comment]
> We are working heavily on performing the update to 4.4 (which does not exhibit the same issue).

Do I read you right: you don't have the issue on 4.4?
The big difference between 3.4 and 4.4 is that services are now private by default.
If confirmed, then I'd be tempted to close as "won't fix", since such graph will quickly disappear with 3.4 entering EOL in a few months.

**Comment by markomitranic** [comment]
> > We are working heavily on performing the update to 4.4 (which does not exhibit the same issue).
> 
> Do I read you right: you don't have the issue on 4.4?
> The big difference between 3.4 and 4.4 is that services are now private by default.
> If confirmed, then I'd be tempted to close as "won't fix", since such graphs will quickly disappear with 3.4 entering EOL in a few months.

Correct, the issue is seen only at 3.4 branch, and still present on the latest 3.4.
Yes, as stated above, i did not expect a fix, since 3.4 will EOL soon, but am very happy to hear your thoughts on the issue.

So in theory, if i understood correctly, if we marked, for example, most of the 1700 services as private, they won't go into linear recursion as seen here, since they will be treated differently?

**Comment by nicolas-grekas** [comment]
Correct: private services can be inlined. This changes the graph and I guess that's where the difference matters here.

**Comment by markomitranic** [comment]
Beautiful, a bit of regex just solved our issues until 4.4 hits us in August. Thank you for the response :)

**Comment by YuriStruszczynski** [comment]
> 
> 
> Beautiful, a bit of regex just solved our issues until 4.4 hits us in August. Thank you for the response :)

Hi, can you please provide your solution in Detail? Im not sure, but we loose like 500ms everytime in checking for $cache->isFresh(). Maybe its related, so i want to try your fix :-)
